In [1]:
import pandas as pd

# Load each data file into a DataFrame
data_file1 = pd.read_csv("C:\\Users\\admin\\Downloads\\data_1.csv\\data_1.csv")
data_file2 = pd.read_csv("C:\\Users\\admin\\Downloads\\data_2.csv\\data_2.csv")
data_file3 = pd.read_csv("C:\\Users\\admin\\Desktop\\python workspace\\finaldataset_3.csv")
data_file4 = pd.read_csv("C:\\Users\\admin\\Downloads\\data_4.csv\\data_4.csv")


In [2]:
# Concatenate the dataframes
merged_data = pd.concat([data_file1, data_file2, data_file3, data_file4], ignore_index=True)


In [3]:
merged_data.drop('YearOfRelease', axis=1, inplace=True)

In [4]:
merged_data

MovieID  CustomerID  Rating            Title
0               1     1488844       3  Dinosaur Planet
1               1      822109       5  Dinosaur Planet
2               1      885013       4  Dinosaur Planet
3               1       30878       4  Dinosaur Planet
4               1      823519       3  Dinosaur Planet
...           ...         ...     ...              ...
99256952    17770     1790158       4     Alien Hunter
99256953    17770     1608708       3     Alien Hunter
99256954    17770      234275       1     Alien Hunter
99256955    17770      255278       4     Alien Hunter
99256956    17770      453585       2     Alien Hunter

[99256957 rows x 4 columns]

In [5]:
# Assuming 'merged data' is your DataFrame and 'Customer_Id' is the column containing user IDs

import pandas as pd

# Define aggregation functions
f = ['count', 'mean']

# Group by 'Customer_Id' and aggregate ratings
dataset_cust_summary = merged_data.groupby('CustomerID')['Rating'].agg(f)

# Convert index to integer
dataset_cust_summary.index = dataset_cust_summary.index.map(int)

# Calculate the threshold for active users (quantile 0.8)
cust_benchmark = round(dataset_cust_summary['count'].quantile(0.8), 0)

# Extract inactive users (users who rate less often)
inactive_users = dataset_cust_summary[dataset_cust_summary['count'] < cust_benchmark].index

print(f'Customer minimum times of review: {cust_benchmark}')

# Create a list of active users (users who rate more often)
active_users = dataset_cust_summary[dataset_cust_summary['count'] >= cust_benchmark].index

# Print the list of active users
print("Active Users (Users who rate more often):")
print(active_users)

# Print the list of inactive users
print("Inactive Users (Users who rate less often):")
print(inactive_users)

num_active_users = len(active_users)
num_inactive_users = len(inactive_users)

print(f'Number of active users: {num_active_users}')
print(f'Number of inactive users: {num_inactive_users}')

Customer minimum times of review: 318.0
Active Users (Users who rate more often):
Index([      6,       7,      79,      97,     134,     188,     195,     199,
           201,     265,
       ...
       2649231, 2649257, 2649267, 2649285, 2649296, 2649308, 2649328, 2649335,
       2649336, 2649378],
      dtype='int64', name='CustomerID', length=96260)
Inactive Users (Users who rate less often):
Index([      8,      10,      25,      33,      42,      59,      83,      87,
            94,     116,
       ...
       2649376, 2649379, 2649384, 2649388, 2649401, 2649404, 2649409, 2649421,
       2649426, 2649429],
      dtype='int64', name='CustomerID', length=383921)
Number of active users: 96260
Number of inactive users: 383921


Customer minimum times of review: 318.0
Number of active users: 96260
Number of inactive users: 383921



users based filtering

In [6]:
# Convert 'CustomerID' column to integer
merged_data['CustomerID'] = merged_data['CustomerID'].astype(int)

# Ensure the active_users list contains integers
active_users = [int(user) for user in active_users]

# Filter the data to exclude inactive users (i.e., keep only active users)
filtered_active_users = merged_data[merged_data['CustomerID'].isin(active_users)]
filtered_inactive_users = merged_data[~merged_data['CustomerID'].isin(active_users)]

# Display the filtered data for active users
print("Active Users' Data:")
print(filtered_active_users)


# Display the filtered data for inactive users
print("\nInactive Users' Data:")
print(filtered_inactive_users)


Active Users' Data:
          MovieID  CustomerID  Rating            Title
0               1     1488844       3  Dinosaur Planet
2               1      885013       4  Dinosaur Planet
3               1       30878       4  Dinosaur Planet
4               1      823519       3  Dinosaur Planet
5               1      893988       3  Dinosaur Planet
...           ...         ...     ...              ...
99256952    17770     1790158       4     Alien Hunter
99256953    17770     1608708       3     Alien Hunter
99256954    17770      234275       1     Alien Hunter
99256955    17770      255278       4     Alien Hunter
99256956    17770      453585       2     Alien Hunter

[63063095 rows x 4 columns]

Inactive Users' Data:
          MovieID  CustomerID  Rating            Title
1               1      822109       5  Dinosaur Planet
6               1      124105       4  Dinosaur Planet
8               1     1842128       4  Dinosaur Planet
9               1     2238063       3  Dinosaur 

In [7]:
# Group the filtered DataFrames by 'CustomerID' and count the number of ratings for each user
active_user_rating_counts = filtered_active_users.groupby('CustomerID').size()
inactive_user_rating_counts = filtered_inactive_users.groupby('CustomerID').size()
 
# Print the number of ratings for active users
print("Number of ratings for active users:")
print(active_user_rating_counts)
 
# Print the number of ratings for inactive users
print("\nNumber of ratings for inactive users:")
print(inactive_user_rating_counts)


Number of ratings for active users:
CustomerID
6          618
7          869
79         787
97         502
134        730
          ... 
2649308    575
2649328    318
2649335    338
2649336    368
2649378    338
Length: 96260, dtype: int64

Number of ratings for inactive users:
CustomerID
8           96
10         257
25          27
33          36
42         126
          ... 
2649404     53
2649409     50
2649421     25
2649426    291
2649429    260
Length: 383921, dtype: int64


318< then inactive , >318 active which is 80%

In [9]:
# Check if customer ID 1 is active
if 6 in active_users:
    print("Customer ID 10 is active.")
else:
    print("Customer ID 10 is inactive.")

Customer ID 10 is active.


In [ ]:
movie based filtering

In [10]:
# Define aggregation functions
f = ['count', 'mean']

# Group by 'MovieID' and aggregate ratings
dataset_movie_summary = filtered_active_users.groupby('MovieID')['Rating'].agg(f)

# Convert index to integer
dataset_movie_summary.index = dataset_movie_summary.index.map(int)

# Calculate the threshold for active movies (quantile 0.8)
movie_benchmark = round(dataset_movie_summary['count'].quantile(0.8), 0)

# Extract inactive movies (movies with fewer ratings)
inactive_movies = dataset_movie_summary[dataset_movie_summary['count'] < movie_benchmark].index

print(f'Movie minimum times of review: {movie_benchmark}')

# Create a list of active movies (movies with more ratings)
active_movies = dataset_movie_summary[dataset_movie_summary['count'] >= movie_benchmark].index

# Print the list of active movies
print("Active Movies (Movies with more ratings):")
print(active_movies)

# Print the list of inactive movies
print("Inactive Movies (Movies with fewer ratings):")
print(inactive_movies)

num_active_movies = len(active_movies)
num_inactive_movies = len(inactive_movies)

print(f'Number of active movies: {num_active_movies}')
print(f'Number of inactive movies: {num_inactive_movies}')


Movie minimum times of review: 3045.0
Active Movies (Movies with more ratings):
Index([    8,    17,    18,    26,    28,    30,    33,    44,    46,    58,
       ...
       17707, 17709, 17724, 17725, 17730, 17743, 17756, 17762, 17764, 17769],
      dtype='int64', name='MovieID', length=3503)
Inactive Movies (Movies with fewer ratings):
Index([    1,     2,     3,     4,     5,     6,     7,     9,    10,    11,
       ...
       17758, 17759, 17760, 17761, 17763, 17765, 17766, 17767, 17768, 17770],
      dtype='int64', name='MovieID', length=14008)
Number of active movies: 3503
Number of inactive movies: 14008


minimum times a movie has been reviewed 3045 for quantile 0.8
3045< active movies <3045 then inactive movies
Number of active movies: 3503
Number of inactive movies: 14008
active movies =most rated movies
WE ARE SEEING NUMBER OF RATINGS not actual ratings (4,5,etc)

In [11]:
# Convert 'MovieID' column to integer
filtered_active_users['MovieID'] = filtered_active_users['MovieID'].astype(int)

# Ensure the active_movies list contains integers
active_movies = [int(movie) for movie in active_movies]

# Filter the data to include only active movies
filtered_active_movies = filtered_active_users[filtered_active_users['MovieID'].isin(active_movies)]

# Filter the data to exclude active movies (i.e., keep only inactive movies)
filtered_inactive_movies = merged_data[~merged_data['MovieID'].isin(active_movies)]

# Display the filtered data for active movies
print("Active Movies' Data:")
print(filtered_active_movies)

# Display the filtered data for inactive movies
print("\nInactive Movies' Data:")
print(filtered_inactive_movies)


C:\Users\admin\AppData\Local\Temp\ipykernel_11520\27302529.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  filtered_active_users['MovieID'] = filtered_active_users['MovieID'].astype(int)


Active Movies' Data:
          MovieID  CustomerID  Rating                       Title
5101            8      785314       1  What the #$*! Do We Know!?
5102            8      243963       3  What the #$*! Do We Know!?
5104            8     1447783       4  What the #$*! Do We Know!?
5108            8     1912665       1  What the #$*! Do We Know!?
5111            8     1744889       1  What the #$*! Do We Know!?
...           ...         ...     ...                         ...
99256027    17769     1428223       4                 The Company
99256030    17769      483107       4                 The Company
99256032    17769       77664       2                 The Company
99256033    17769     1922916       3                 The Company
99256035    17769     1704416       2                 The Company

[55541107 rows x 4 columns]

Inactive Movies' Data:
          MovieID  CustomerID  Rating            Title
0               1     1488844       3  Dinosaur Planet
1               1      8

In [12]:
# Group the filtered DataFrames by 'MovieID' and count the number of ratings for each movie
active_movie_rating_counts = filtered_active_movies.groupby('MovieID').size()
inactive_movie_rating_counts = filtered_inactive_movies.groupby('MovieID').size()

# Print the number of ratings for active movies
print("Number of ratings for active movies:")
print(active_movie_rating_counts)

# Print the number of ratings for inactive movies
print("\nNumber of ratings for inactive movies:")
print(inactive_movie_rating_counts)


Number of ratings for active movies:
MovieID
8         7054
17        3160
18        8546
26        3368
28       27078
         ...  
17743     4097
17756     5891
17762    24940
17764    43673
17769     3867
Length: 3503, dtype: int64

Number of ratings for inactive movies:
MovieID
1         547
2         145
3        2012
4         142
5        1140
         ... 
17765     313
17766     605
17767     204
17768    1362
17770     921
Length: 14008, dtype: int64


In [27]:
from IPython.display import FileLink

# Convert filtered_active_movies DataFrame to CSV and download
filtered_active_movies.to_csv('clean_final.csv', index=False)
display(FileLink('clean_final.csv'))


c:\Users\admin\Desktop\python workspace\clean_final.csv

In [ ]:
# Check if movie ID 9211 is active
if 9224 in active_movies:
    print("Movie ID 9211 is active- most rated.")
else:
    print("Movie ID 9211 is inactive-not most rated.")

Movie ID 9211 is active- most rated.


collaborative filtering

In [1]:
import pandas as pd
clean_data= pd.read_csv("C:\\Users\\admin\\Desktop\\python workspace\\clean_final.csv")

In [56]:
user_movie_interact = clean_data.pivot_table(index = 'Title', columns = 'CustomerID', values = 'Rating')
user_movie_interact.fillna(0, inplace = True)   #filling nan values with 0
user_movie_interact

CustomerID                                   6        7        79       \
Title                                                                    
...And Justice for All                           0.0      0.0      0.0   
10                                               0.0      0.0      0.0   
10 Things I Hate About You                       0.0      0.0      0.0   
100 Girls                                        0.0      0.0      0.0   
101 Dalmatians II: Patch's London Adventure      0.0      0.0      3.0   
...                                              ...      ...      ...   
Yours                                            0.0      0.0      5.0   
Zelig                                            0.0      0.0      0.0   
Zero Effect                                      0.0      0.0      0.0   
Zoolander                                        3.0      2.0      4.0   
Zulu                                             0.0      0.0      0.0   

CustomerID                                   97       134      188      \
Title                                                                    
...And Justice for All                           0.0      0.0      0.0   
10                                               0.0      0.0      0.0   
10 Things I Hate About You                       0.0      5.0      0.0   
100 Girls                                        0.0      0.0      0.0   
101 Dalmatians II: Patch's London Adventure      0.0      0.0      0.0   
...                                              ...      ...      ...   
Yours                                            0.0      0.0      0.0   
Zelig                                            0.0      0.0      0.0   
Zero Effect                                      0.0      0.0      0.0   
Zoolander                                        0.0      0.0      3.0   
Zulu                                             0.0      0.0      0.0   

CustomerID                                   195      199      201      \
Title                                                                    
...And Justice for All                           0.0      0.0      0.0   
10                                               0.0      0.0      0.0   
10 Things I Hate About You                       0.0      0.0      4.0   
100 Girls                                        0.0      0.0      0.0   
101 Dalmatians II: Patch's London Adventure      0.0      0.0      0.0   
...                                              ...      ...      ...   
Yours                                            0.0      0.0      0.0   
Zelig                                            0.0      0.0      0.0   
Zero Effect                                      0.0      0.0      0.0   
Zoolander                                        2.0      5.0      3.0   
Zulu                                             0.0      0.0      0.0   

CustomerID                                   265      ...  2649231  2649257  \
Title                                                 ...                     
...And Justice for All                           0.0  ...      0.0      0.0   
10                                               0.0  ...      0.0      0.0   
10 Things I Hate About You                       0.0  ...      3.0      4.0   
100 Girls                                        0.0  ...      0.0      0.0   
101 Dalmatians II: Patch's London Adventure      0.0  ...      0.0      0.0   
...                                              ...  ...      ...      ...   
Yours                                            0.0  ...      0.0      0.0   
Zelig                                            0.0  ...      0.0      0.0   
Zero Effect                                      0.0  ...      0.0      0.0   
Zoolander                                        3.0  ...      0.0      0.0   
Zulu                                             0.0  ...      0.0      0.0   

CustomerID                                   2649267  2649285  2649296  \
Title    

In [57]:
pip install scikit-learn

^C
Note: you may need to restart the kernel to use updated packages.


In [58]:
from sklearn.metrics.pairwise import cosine_similarity
similarity_score = cosine_similarity(user_movie_interact)
similarity_score

array([[1.        , 0.22911379, 0.10244852, ..., 0.11895454, 0.12917896,
        0.16226369],
       [0.22911379, 1.        , 0.13371459, ..., 0.12734618, 0.14474661,
        0.13730845],
       [0.10244852, 0.13371459, 1.        , ..., 0.14015507, 0.44803133,
        0.06838905],
       ...,
       [0.11895454, 0.12734618, 0.14015507, ..., 1.        , 0.2211853 ,
        0.07922016],
       [0.12917896, 0.14474661, 0.44803133, ..., 0.2211853 , 1.        ,
        0.09392063],
       [0.16226369, 0.13730845, 0.06838905, ..., 0.07922016, 0.09392063,
        1.        ]])

In [59]:
similarity_score.shape
#converting to a dataframe
similar_movie_df = pd.DataFrame(similarity_score, index = user_movie_interact.index, columns = user_movie_interact.index)
similar_movie_df

Title                                        ...And Justice for All        10  \
Title                                                                           
...And Justice for All                                     1.000000  0.229114   
10                                                         0.229114  1.000000   
10 Things I Hate About You                                 0.102449  0.133715   
100 Girls                                                  0.052211  0.086965   
101 Dalmatians II: Patch's London Adventure                0.077694  0.106595   
...                                                             ...       ...   
Yours                                                      0.149367  0.152818   
Zelig                                                      0.153525  0.135174   
Zero Effect                                                0.118955  0.127346   
Zoolander                                                  0.129179  0.144747   
Zulu                                                       0.162264  0.137308   

Title                                        10 Things I Hate About You  \
Title                                                                     
...And Justice for All                                         0.102449   
10                                                             0.133715   
10 Things I Hate About You                                     1.000000   
100 Girls                                                      0.225043   
101 Dalmatians II: Patch's London Adventure                    0.196959   
...                                                                 ...   
Yours                                                          0.210781   
Zelig                                                          0.067691   
Zero Effect                                                    0.140155   
Zoolander                                                      0.448031   
Zulu                                                           0.068389   

Title                                        100 Girls  \
Title                                                    
...And Justice for All                        0.052211   
10                                            0.086965   
10 Things I Hate About You                    0.225043   
100 Girls                                     1.000000   
101 Dalmatians II: Patch's London Adventure   0.071721   
...                                                ...   
Yours                                         0.055919   
Zelig                                         0.029225   
Zero Effect                                   0.077481   
Zoolander                                     0.168800   
Zulu                                          0.034700   

Title                                        101 Dalmatians II: Patch's London Adventure  \
Title                                                                                      
...And Justice for All                                                          0.077694   
10                                                                              0.106595   
10 Things I Hate About You                                                      0.196959   
100 Girls                                                                       0.071721   
101 Dalmatians II: Patch's London Adventure                                     1.000000   
...                                                                                  ...   
Yours                                                                           0.175015   
Zelig                                                                           0.023766   
Zero Effect                                                                     0.036911   
Zoolander                                                                       0.129680   
Zulu                                                                            0.063512   

Title             

In [39]:
# Defining column names
movie_titles=pd.read_csv("C:\\Users\\admin\\Downloads\\movie_titles_final.csv",encoding='ISO-8859-1',usecols = ['MovieID', 'YearOfRelease', 'Title'],on_bad_lines ='skip')



In [40]:
movie_titles['MovieID'] = movie_titles['MovieID'].astype('int16')

In [41]:
clean_data['MovieID'] = clean_data['MovieID'].astype('int16')

In [42]:
data_num_ratings = clean_data.copy(deep = True)

In [43]:
data_num_ratings['num_ratings'] = data_num_ratings.groupby('MovieID')['Rating'].transform('mean')

In [44]:
data_num_ratings = data_num_ratings.sort_values(by = ["Rating",'num_ratings'], ascending = [False,False])

In [45]:
data_num_ratings

MovieID  CustomerID  Rating  \
46555171    14961      577397       5   
46555172    14961      185150       5   
46555173    14961     2639583       5   
46555174    14961     1185389       5   
46555175    14961      951709       5   
...           ...         ...     ...   
38464722    12517     2506617       1   
38464725    12517       63786       1   
38464729    12517      926461       1   
38464738    12517     2646249       1   
38464739    12517      823628       1   

                                                      Title  num_ratings  
46555171  Lord of the Rings: The Return of the King: Ext...     4.712575  
46555172  Lord of the Rings: The Return of the King: Ext...     4.712575  
46555173  Lord of the Rings: The Return of the King: Ext...     4.712575  
46555174  Lord of the Rings: The Return of the King: Ext...     4.712575  
46555175  Lord of the Rings: The Return of the King: Ext...     4.712575  
...                                                     ...          ...  
38464722                                              Gigli     1.939198  
38464725                                              Gigli     1.939198  
38464729                                              Gigli     1.939198  
38464738                                              Gigli     1.939198  
38464739                                              Gigli     1.939198  

[55541107 rows x 5 columns]

In [46]:
unique_ratings = data_num_ratings.drop_duplicates(subset=['Title', 'num_ratings'])

In [47]:
unique_ratings

MovieID  CustomerID  Rating  \
46555171    14961      577397       5   
22497591     7230     2639583       5   
22043810     7057     2639583       5   
10103901     3456     1091832       5   
5986737      2102     1987434       5   
...           ...         ...     ...   
50322215    16146     1270549       5   
25121732     8101      181323       5   
37070199    12171      780597       5   
8699356      3021      249052       5   
38458293    12517     2187374       5   

                                                      Title  num_ratings  
46555171  Lord of the Rings: The Return of the King: Ext...     4.712575  
22497591  The Lord of the Rings: The Fellowship of the R...     4.706620  
22043810  Lord of the Rings: The Two Towers: Extended Ed...     4.693892  
10103901                                     Lost: Season 1     4.659735  
5986737                              The Simpsons: Season 6     4.584646  
...                                                     ...          ...  
50322215                                    The Real Cancun     2.163748  
25121732                                       Full Frontal     2.104738  
37070199                                  Battlefield Earth     2.101558  
8699356                                   House of the Dead     1.964487  
38458293                                              Gigli     1.939198  

[3503 rows x 5 columns]

num ratings= mean of rating, LL 5 ratings would come so need to drop duplicates

In [34]:
data1_num_ratings = data_num_ratings.drop_duplicates(subset =['Title', 'MovieID'],keep = "first")

KeyError: Index(['MovieID'], dtype='object')

In [37]:
data_num_ratings.drop(columns=['CustomerID', 'MovieID'], inplace=True)

KeyError: "['CustomerID', 'MovieID'] not found in axis"

In [26]:
movie_titles.head()

MovieID  YearOfRelease                         Title
0        1         2003.0               Dinosaur Planet
1        2         2004.0    Isle of Man TT 2004 Review
2        3         1997.0                     Character
3        4         1994.0  Paula Abdul's Get Up & Dance
4        5         2004.0      The Rise and Fall of ECW

In [70]:
def recommend(movie_title, rating=5):
    if movie_title in similar_movie_df.index:
        score = similar_movie_df[movie_title]*(rating-2.5)
        score = score.sort_values(ascending = False)[1:6]    #recommend top 5 movies

        return score

    else:
        top_movies = unique_ratings[:6]
        print(f'"{movie_title}" not found in the database')
        print("Watch these instead:")
        return top_movies[['Title', 'num_ratings']]

In [71]:
recommend("Legally Blonde")

Title
Miss Congeniality           1.966225
Sweet Home Alabama          1.940233
Pretty Woman                1.903660
My Big Fat Greek Wedding    1.875781
Erin Brockovich             1.853188
Name: Legally Blonde, dtype: float64

In [72]:
recommend("Sholay")

"Sholay" not found in the database
Watch these instead:


Title  num_ratings
46555171  Lord of the Rings: The Return of the King: Ext...     4.712575
22497591  The Lord of the Rings: The Fellowship of the R...     4.706620
22043810  Lord of the Rings: The Two Towers: Extended Ed...     4.693892
10103901                                     Lost: Season 1     4.659735
5986737                              The Simpsons: Season 6     4.584646
44864011          The Shawshank Redemption: Special Edition     4.556081